### ISSUE TITLE
[UX] Field UI - clean up the field type selection

### ISSUE THREAD

**Issue by jenlampton** [backdrop/backdrop-issues#1041]
There are a few problems that users face when adding fields.  The first is the problem of selecting a type of field.
![screen shot 2015-06-27 at 1 39 32 pm](https://cloud.githubusercontent.com/assets/397895/8393568/4f945db6-1cd2-11e5-9625-d3acb97ceb7c.png)
Possible suggestions for cleaning up this list:
* [ ] Rename List (type) items into Type (list) for improved scannability

**Comment by docwilmot** [comment]
> Move the widget column to the right of the "Field type" column

If what you actually meant to write was "Move the widget column to the **left** of the "Field type" column" then I love number two (and by extension then number three): the idea here is the user would more likely have an idea of what type of element (widget) she wants on the page (a checkbox, a select list, a textfield) so present this choice first, then allow him to choose what type of data to store in it afterwards. Sounds like a great idea. 

Number one I'd simplify even further, to just "Decimal number" instead of "Decimal (number)" as proposed.

**Comment by quicksketch** [comment]
>  the idea here is the user would more likely have an idea of what type of element (widget) she wants on the page (a checkbox, a select list, a textfield) so present this choice first, then allow him to choose what type of data to store in it afterwards. Sounds like a great idea. 

I'm pretty sure that's what was meant here. Completely agreed. Show the widget select first and the data type second.

**Comment by MrHaroldA** [comment]
And what about lists? Can't we just add cardinality to a type-field instead?

A single checkbox could just as well be a list with just one option.

**Comment by klonos** [comment]
This idea of having the user select a widget first and then the data type reminds me of how [Hierarchical Select](https://www.drupal.org/project/hierarchical_select) works. Was that what you were thinking?

**Comment by klonos** [comment]
I'm not sure the PR that @hellomrcat has filed goes towards the right direction. Maybe because the issue summary still suggests changing the field type from ```widget_type (data_type)``` to ```data_type (widget_type)``` which I think would be a totally wrong move. It would revert the changes made in https://www.drupal.org/node/2072533 in order to "group" data types together for better scanability.and we'd return to a list like this (alphabetical order):

Boolean
Decimal number
File
Float list
Float number
Image
Integer list
Integer number
Long text
Long text (with summary)
Short text
Term reference
Text list

...and while by pure (alphabetical) luck the text options are grouped together, the options for numbers are all over the place.

So we have three suggestions here:

1. change ```widget_type (data_type)``` to ```data_type (widget_type)```.
2. move the widget column to the **left** of the "Field type" column.
3. rename the "Field type" column to "Data type"

I think 1 is a no-no, 2 a :+1: and 3 not sure.

When it comes to 3, personally I'd like to clarify things a bit more and have these column titles be "Field widget" and "Field data type"

**Comment by quicksketch** [comment]
Yeah, the PR needs some work. Here's an annotated screenshot if that helps.
![selection_005](https://cloud.githubusercontent.com/assets/100206/9047077/2ff26a0a-39e4-11e5-9fd1-4da2fd45026b.png)

**Comment by jenlampton** [comment]
> ...and while by pure (alphabetical) luck the text options are grouped together, the options for numbers and lists are all over the place.

This isn't luck, it's grouped sanely based on the way we name things. If numbers are all over the place, then we should change the names for numbers (and lists too) like so: 

Boolean
File
Image
Number: Decimal
Number: Float
Number: Float (list)
Number: Integer 
Number: Integer (list)
Long text
Long text (with summary)
Short text
Term reference
Text list

**Comment by philsward** [comment]
Not opposed to what @quicksketch suggested. I think I would have to play with it a bit to see if it would be more (or less) confusing.

> This isn't luck, it's grouped sanely based on the way we name things. If numbers are all over the place, then we should change the names for numbers (and lists, too) like so:

I like where you're going with this @jenlampton 

Disclaimer: I'm fairly new to BDCMS

Several things from my perspective:

1) When I saw the List (Integer) & List (Float) all I could think was: "Huh? What are they trying to accomplish?" It wasn't until reading this thread and viewing Jen's proposal that it made a bit more sense: "A list of numbers".

2) I've always hated the Numbers field nomenclature Integer, Float and Decimal from Drupal. I honestly did horrible in math and couldn't tell you what a float or integer is without actually looking up the definition. It is rare that I use the number field and to be honest, unless I need to lock the field to an actual number, I've found it is just easier to use a text field to represent numbers.

3) I think it might be worth re-doing how the text fields are presented. BD has some widgets that I don't remember being present with Drupal: Text Field and Text Area specifically with the short text. It honestly threw me for a loop at first "Text area with a text field? Why?". I later realized there might be some use for showing a long text area within the char limit of the text field, so it does make sense but from a UX standpoint it's kind of confusing to see both widgets. Maybe this is a Drupal mindset I need to overcome.

One thought is to bring this into the settings of the field itself and re-do the widget to dynamically accept the choice on the settings page. Best example I can give is "Plain Text vs Filtered Text". Add a radio button like "Allow one row of text vs Allow multiple rows of text".

Another thought on all of this, is to collect and dynamically set how the data is saved on the field creation itself.

So instead of Number (decimal) Number (float) Number (Integer), we are asked at creation when adding the [Number] field :

What type of number do you want to store?
Decimal: A number with a decimal point, for example 2.00.
Float: A decimal number that can either be a positive or negative number for example 5.5 or -0.0023
Integer: A positive number without any decimals for example 5, 100 or 65234
Phone: A Phone Number

---

Going this route would honestly require some work, but it would reduce the clutter when viewing the field type select list. It would also be less confusing to new folks who know they want a number, or maybe a text field, but don't yet understand the difference between a long text and short text, or the different types of numbers.

This does create an additional page for collecting info which is a pain-point for additional steps (the less the better) but it would allow a much better explanation on that data type collection page to explain what each data type is. We don't get this now with a dropdown of Number (float) or Text (short).

**Comment by klonos** [comment]
Thanks for chiming in @philsward 

>I've always hated the Numbers field nomenclature Integer, Float and Decimal from Drupal. I honestly did horrible in math and couldn't tell you what a float or integer is without actually looking up the definition. 

Then, you may also be interested in #4204 re that 😉 

> It is rare that I use the number field and to be honest, unless I need to lock the field to an actual number, I've found it is just easier to use a text field to represent numbers.

Yup, number validation is one of the benefits of using number fields. Other benefits include better support in mobile (only the numeric keyboard is shown), being able to do calculations without needing to convert strings to numbers, as well as internationalization support (swapping commas/dots for thousands/decimals etc.).

There's also #779 which I'd really like us eventually to get to.
